packages/bitcore-lib/lib/script/interpreter.js

```diff
@@ -610,54 +965,104 @@ Interpreter.prototype.checkLockTime = function(nLockTime) {
-    // simple numeric one.
-    if (nSequenceMasked.gt(txToSequenceMasked)) {
-        return false;
+  // There are two kinds of nSequence: lock-by-blockheight and
+  // lock-by-blocktime, distinguished by whether nSequenceMasked <
+  // CTxIn::SEQUENCE_LOCKTIME_TYPE_FLAG.
+  //
+  // We want to compare apples to apples, so fail the script unless the type
+  // of nSequenceMasked being tested is the same as the nSequenceMasked in the
+  // transaction.
+  var SEQUENCE_LOCKTIME_TYPE_FLAG_BN = new BN(Interpreter.SEQUENCE_LOCKTIME_TYPE_FLAG);
+  
+  if (!((txToSequenceMasked.lt(SEQUENCE_LOCKTIME_TYPE_FLAG_BN)  &&
+          nSequenceMasked.lt(SEQUENCE_LOCKTIME_TYPE_FLAG_BN)) ||
+        (txToSequenceMasked.gte(SEQUENCE_LOCKTIME_TYPE_FLAG_BN) &&
+          nSequenceMasked.gte(SEQUENCE_LOCKTIME_TYPE_FLAG_BN)))) {
+    return false;
+  }
+
+  // Now that we know we're comparing apples-to-apples, the comparison is a
+  // simple numeric one.
+  if (nSequenceMasked.gt(txToSequenceMasked)) {
+    return false;
+  }
+  return true;
+}
+
+
+Interpreter.computeTapleafHash = function(leafVersion, scriptBuf) {
+  const tagWriter = TaggedHash.TAPLEAF;
+  tagWriter.writeUInt8(leafVersion);
+  tagWriter.writeVarintNum(scriptBuf.length);
+  tagWriter.write(scriptBuf);
+  return tagWriter.finalize();
+};
+
+
+Interpreter.computeTaprootMerkleRoot = function(control, tapleafHash) {
+  const pathLen = (control.length - Interpreter.TAPROOT_CONTROL_BASE_SIZE) / Interpreter.TAPROOT_CONTROL_NODE_SIZE;
+  let k = tapleafHash;
+  for (let i = 0; i < pathLen; ++i) {
+    const tagWriter = TaggedHash.TAPBRANCH;
+    const start = Interpreter.TAPROOT_CONTROL_BASE_SIZE + Interpreter.TAPROOT_CONTROL_NODE_SIZE * i;
+    const node = control.slice(start, start + Interpreter.TAPROOT_CONTROL_NODE_SIZE);
+    if (Buffer.compare(k, node) === -1) {
+      tagWriter.write(k);
+      tagWriter.write(node);
+    } else {
+      tagWriter.write(node);
+      tagWriter.write(k);
     }
-    return true;
+    k = tagWriter.finalize();
   }
+  return k;
+};
+
+
+Interpreter.verifyTaprootCommitment = function(control, program, tapleafHash) {
+  $.checkArgument(control.length >= Interpreter.TAPROOT_CONTROL_BASE_SIZE, 'control too short');
+  $.checkArgument(program.length >= 32, 'program is too short');
+
+  try {
+    //! The internal pubkey (x-only, so no Y coordinate parity).
+    const p = PublicKey.fromX(false, control.slice(1, Interpreter.TAPROOT_CONTROL_BASE_SIZE));
+    //! The output pubkey (taken from the scriptPubKey).
+    const q = PublicKey.fromX(false, program);
+    // Compute the Merkle root from the leaf and the provided path.
+    const merkleRoot = Interpreter.computeTaprootMerkleRoot(control, tapleafHash);
+    // Verify that the output pubkey matches the tweaked internal pubkey, after correcting for parity.
+    return q.checkTapTweak(p, merkleRoot, control);
+  } catch (err) {
+    // TODO: set this.errstr here?
```

Comment:
ping in case you want to address this TODO

Reply:
The only place that uses this fn sets `this.errstr` if false. Removed comment.